packages/reakit/src/Form/__tests__/index-test.tsx

```diff
@@ -37,13 +37,14 @@ test(
     const input = getByLabelText("input");
     expect(onValidate).not.toHaveBeenCalled();
     fireEvent.change(input, { target: { value: "a" } });
+    await Promise.resolve();
```

Comment:
Why is this necessary? Looks like validate on change doesn't work when testing on the browser.

Reply:
Because of how I implemented the validation function, `onValidate` is only called after the main thread is free. I can refactor `Promise.resolve().then` into `new Promise(resolve =>` and this will be unnecessary, should I do it?

Reply:
I think the issue with it not working on the browser was caused by the `if (messages) { dispatch(/* ... */) }` you reported. I'll test it

Reply:
Tested it and it seems to be working, but if you prefer it to be sync, just tell me.

Reply:
Nice! What if we did `Promise.resolve(onValidate()).then(...)` instead? It must work on these scenarios:
- `onValidate` may be undefined;
- `onValidate` may be a normal function that returns undefined or an object of messages;
- `onValidate` may be a normal function that throws an object of errors;
- `onValidate` may be a promise that returns undefined of an object of messages;
- `onValidate` may be a promise that throws an object of errors.

Reply:
I did it like that:
```ts
const validate = React.useCallback(
  (vals = state.values) =>
    new Promise(resolve => {
      if (onValidate) {
        const response = onValidate(filterAllEmpty(vals));
        if (isPromise(response)) {
          dispatch({ type: "startValidate" });
        }

        resolve(
          Promise.resolve(response).then(messages => {
            dispatch({ type: "endValidate", messages });
            return messages;
          })
        );
      } else {
        resolve(undefined);
      }
    }).catch(errors => {
      dispatch({ type: "endValidate", errors });
      throw errors;
    }),
  [state.values, onValidate]
);
```

Reply:
As the function passed to `Promise`'s constructor is run synchronously, everythng is sync up to the `resolve` call.

Reply:
I pushed it for you to see: https://github.com/reakit/reakit/pull/365/commits/4c283e1538002cdc84f94ee37e05ab9845cf1485

Reply:
Cool! Have a meeting now. Will take a look tonight. Thank you ❤️